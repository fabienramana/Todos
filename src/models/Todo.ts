export default interface Todo{
    content: string,
    completed: boolean
}