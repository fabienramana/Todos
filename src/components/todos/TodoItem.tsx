import React, { useEffect, useRef, useState } from "react";
import Todo from "../../models/Todo"

type TodoItemProps = {
    todo: Todo,
    removeTodo: (todo: Todo) => void,
    changeStatusOfTodo: (todoStatus: boolean, index: number) => void,
    changeContentOfTodo: (todoContent: string, index: number) => void
}

export default function TodoItem({ todo, removeTodo, changeStatusOfTodo, changeContentOfTodo }: TodoItemProps): JSX.Element{

    const inputModifierRef = useRef<HTMLInputElement>(null);
    const [editMode, setEditMode] = useState<boolean>(false);
    const [modifiedTodoContent, setModifiedTodoContent] = useState<string>(todo.content)


    const handleCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        console.log(event.target.checked)
        console.log(todo.id)
        changeStatusOfTodo(event.target.checked, todo.id);     
    };

    const handleChangeValue = (event: React.ChangeEvent<HTMLInputElement>) => {
        setModifiedTodoContent(event.target.value);
    };
    
    const handleClick = (event: React.MouseEvent<HTMLLabelElement>) => {
        if(event.detail === 2){
            setEditMode(!editMode)
        }
    }

    function handleKeyPress(e: React.KeyboardEvent<HTMLInputElement>){
        
         if(e.key === 'Enter' && modifiedTodoContent.trim() !== ""){
            changeContentOfTodo(modifiedTodoContent.trim(), todo.id)
            setEditMode(!editMode)
         }
        else if (e.key === 'Enter' && modifiedTodoContent.trim() === ""){
            removeTodo(todo);
            setEditMode(!editMode)
         }

        else if (e.key === 'Escape'){
            setEditMode(!editMode)
         }
    }
    
     useEffect(() => {
        if(editMode && inputModifierRef.current){
            inputModifierRef.current.focus();
        }
    }, [editMode]) 


    return(
        
        <li className={todo.completed ? editMode ? "completed editing" : "completed"  : editMode ? "editing" : "" }>
            <div className="view">
                <input className="toggle" aria-label="changeStatus" type="checkbox" onChange={handleCheckboxChange} checked={todo.completed}/>
                <label onClick={handleClick}>{todo.content}</label>
                <button className="destroy" aria-label="delete" onClick={() => removeTodo(todo)}></button>
            </div>
            <input ref={inputModifierRef} className="edit"
                   onChange={handleChangeValue}
                   onKeyDown={handleKeyPress} 
                   value={modifiedTodoContent}/> 
        </li>
    )
}