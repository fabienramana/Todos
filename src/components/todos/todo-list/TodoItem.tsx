import React, { useEffect, useRef, useState } from "react";
import Todo from "../../../models/Todo"

type TodoItemProps = {
    todoTitle: string,
    todoStatus: boolean,
    removeTodo: () => void,
    changeTodo:(todo: Partial<Todo>) => void
    setEditMode: () => void,
    resetEditMode: () => void
}

export default function TodoItem({ todoTitle, todoStatus, removeTodo, changeTodo, setEditMode, resetEditMode  }: TodoItemProps){

    const inputModifierRef = useRef<HTMLInputElement>(null);
    const [modifiedTodoContent, setModifiedTodoContent] = useState<string>(todoTitle)


    const handleCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        const todoStatus = {
            completed: event.target.checked
        }
        changeTodo(todoStatus);     
    };

    const handleChangeValue = (event: React.ChangeEvent<HTMLInputElement>) => {
        setModifiedTodoContent(event.target.value);
    };
    
    const handleClick = (event: React.MouseEvent<HTMLLabelElement>) => {
        setEditMode()
    }

    function handleKeyPress(e: React.KeyboardEvent<HTMLInputElement>){
        const normalizedInput = modifiedTodoContent.trim();

         if(e.key === 'Enter' && normalizedInput !== ""){
            const todoTitleToChange = {title: normalizedInput}
            changeTodo(todoTitleToChange)
            resetEditMode()
         }
        
         if (e.key === 'Enter' && normalizedInput === ""){
            removeTodo();
            resetEditMode()
         }

         if (e.key === 'Escape'){
            setModifiedTodoContent(todoTitle)
            resetEditMode()
         }
    }
    
      useEffect(() => {
        if(inputModifierRef.current){
            inputModifierRef.current.focus();
        }
    })  


    return(
        
        <div>
            <div className="view">
                <input className="toggle" aria-label="changeStatus" type="checkbox" onChange={handleCheckboxChange} checked={todoStatus}/>
                <label aria-label="title" onDoubleClick={handleClick}>{todoTitle}</label>
                <button className="destroy" aria-label="delete" onClick={() => removeTodo()}></button>
            </div>
            <input ref={inputModifierRef} className="edit"
                   onChange={handleChangeValue}
                   onKeyDown={handleKeyPress}
                   value={modifiedTodoContent}/> 
        </div>
    )
}