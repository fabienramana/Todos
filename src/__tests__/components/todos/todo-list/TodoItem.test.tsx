import { render, screen, fireEvent } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import TodoItem from "../../../../components/todos/todo-list/TodoItem"
import Todo from "../../../../models/Todo"


describe('TodoItem component', () => {
    const todo: Todo ={
      id:1,
      title: "Soccer match at 3 pm",
      completed: false
    }
  
    test('should display todo title', () => {
  
      const removeTodo = () => {}
      const changeTodo = () => {}
      const setEditMode = () => {}
      const resetEditMode = () => {}
  
      const {asFragment} = render(<TodoItem todoTitle={todo.title}
              todoStatus={todo.completed}
              removeTodo={removeTodo}
              changeTodo={changeTodo}
              setEditMode={setEditMode}
              resetEditMode={resetEditMode}/>)
  
      expect(screen.getByText(todo.title)).toBeInTheDocument();
      expect(asFragment()).toMatchSnapshot();
    })
  
    test("should call 'removeTodo' if clicked on delete button", ()=> {
  
      const removeTodo = jest.fn()
      const changeTodo = () => {}
      const setEditMode = () => {}
      const resetEditMode = () => {}
  
      render(<TodoItem todoTitle={todo.title}
              todoStatus={todo.completed}
              removeTodo={removeTodo}
              changeTodo={changeTodo}
              setEditMode={setEditMode}
              resetEditMode={resetEditMode}/>)
  
      fireEvent.click(screen.getByRole('button', {
        name: /delete/
      }));
  
      expect(removeTodo).toBeCalledTimes(1);
      expect(removeTodo).toBeCalledWith()
    })
  
    test("should call changeTodo with a status if checkbox is toggled once", () => {
  
      const removeTodo = () => {}
      const changeTodo = jest.fn()
      const setEditMode = () => {}
      const resetEditMode = () => {}
  
      render(<TodoItem todoTitle={todo.title}
              todoStatus={todo.completed}
              removeTodo={removeTodo}
              changeTodo={changeTodo}
              setEditMode={setEditMode}
              resetEditMode={resetEditMode}/>)

      fireEvent.click(screen.getByRole('checkbox', {
        name: /changeStatus/
      }));
      
      expect(changeTodo).toHaveBeenCalledTimes(1)
      expect(changeTodo).toHaveBeenCalledWith({completed: !todo.completed})
    })


    test("should call 'setEditMode' if label is double clicked", () => {
  
      const removeTodo = () => {}
      const changeTodo = () => {}
      const setEditMode = jest.fn()
      const resetEditMode = () => {}
  
      render(<TodoItem todoTitle={todo.title}
              todoStatus={todo.completed}
              removeTodo={removeTodo}
              changeTodo={changeTodo}
              setEditMode={setEditMode}
              resetEditMode={resetEditMode}/>)

      fireEvent.doubleClick(screen.getByLabelText("title"));
      
      expect(setEditMode).toHaveBeenCalledTimes(1)
    })


    test("should call 'changeTodo' with modified string in input is changed, then enter is pressed", () => {
  
      const removeTodo = () => {}
      const changeTodo = jest.fn()
      const setEditMode = () => {}
      const resetEditMode = () => {}
  
      render(<TodoItem todoTitle={todo.title}
              todoStatus={todo.completed}
              removeTodo={removeTodo}
              changeTodo={changeTodo}
              setEditMode={setEditMode}
              resetEditMode={resetEditMode}/>)

      const input = screen.getByRole("textbox")

      fireEvent.change(input, {target: {value: ''}})
      const modifiedString = `${todo.title}aaa`

      userEvent.type(input, `${modifiedString}{Enter}`);
      
      expect(changeTodo).toHaveBeenCalledTimes(1)
      expect(changeTodo).toHaveBeenCalledWith({title: `${modifiedString}`})
    })

    test("should call 'removeTodo' if input is changed to blank, then enter is pressed", () => {
  
      const removeTodo = jest.fn()
      const changeTodo = () => {}
      const setEditMode = () => {}
      const resetEditMode = () => {}
  
      render(<TodoItem todoTitle={todo.title}
              todoStatus={todo.completed}
              removeTodo={removeTodo}
              changeTodo={changeTodo}
              setEditMode={setEditMode}
              resetEditMode={resetEditMode}/>)

      const input = screen.getByRole("textbox")
      fireEvent.change(input, {target: {value: ''}})
      userEvent.type(input, `{Enter}`);
      
      expect(removeTodo).toHaveBeenCalledTimes(1)
      expect(removeTodo).toHaveBeenCalledWith()
    })

    test("should call 'removeTodo' if input is filled with whitespaces blank, then enter is pressed", () => {
  
      const removeTodo = jest.fn()
      const changeTodo = () => {}
      const setEditMode = () => {}
      const resetEditMode = () => {}
  
      render(<TodoItem todoTitle={todo.title}
              todoStatus={todo.completed}
              removeTodo={removeTodo}
              changeTodo={changeTodo}
              setEditMode={setEditMode}
              resetEditMode={resetEditMode}/>)

      const input = screen.getByRole("textbox")
      fireEvent.change(input, {target: {value: ''}})
      userEvent.type(input, `{Enter}`);
      
      expect(removeTodo).toHaveBeenCalledTimes(1)
      expect(removeTodo).toHaveBeenCalledWith()
    })
  })