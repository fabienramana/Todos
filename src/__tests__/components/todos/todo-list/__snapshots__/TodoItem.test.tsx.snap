// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem component should display todo title 1`] = `
<DocumentFragment>
  <div>
    <div
      class="view"
    >
      <input
        aria-label="changeStatus"
        class="toggle"
        type="checkbox"
      />
      <label
        aria-label="title"
      >
        Soccer match at 3 pm
      </label>
      <button
        aria-label="delete"
        class="destroy"
      />
    </div>
    <input
      class="edit"
      value="Soccer match at 3 pm"
    />
  </div>
</DocumentFragment>
`;
